sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -249,16 +330,74 @@
   }
 
   /**
+   * Creates an uninitialized {@link Read} {@link PTransform}, using a Kafka {@link Deserializer}
+   * based on {@link AvroCoder}. The coder reads Avro data directly without using an Avro object
+   * container.
+   */
+  @SuppressWarnings("unchecked")
+  public static <K, V> Read<K, V> fromAvro(Class<K> keyClass, Class<V> valueClass) {
+    // Kafka constructs deserializers directly. Pass coder through consumer
+    // configuration.
+    ImmutableMap.Builder<String, Object> builder = new ImmutableMap.Builder<>();
+    Map<String, Object> config = builder
+            .putAll(Read.DEFAULT_CONSUMER_PROPERTIES)
+            .put(CoderBasedKafkaDeserializer.configForKeyDeserializer(), AvroCoder.of(keyClass))
+            .put(CoderBasedKafkaDeserializer.configForValueDeserializer(), AvroCoder.of(valueClass))
+            .build();
+
+    CoderBasedKafkaDeserializer<K> keyDeserializer = new CoderBasedKafkaDeserializer<K>();
+    CoderBasedKafkaDeserializer<V> valueDeserializer = new CoderBasedKafkaDeserializer<V>();
+
+    return new AutoValue_KafkaIO_Read.Builder<K, V>()
+            .setTopics(new ArrayList<String>())
+            .setTopicPartitions(new ArrayList<TopicPartition>())
+            .setKeyCoder(AvroCoder.of(keyClass))
+            .setValueCoder(AvroCoder.of(valueClass))
+            .setKeyDeserializer((Class<Deserializer<K>>) (Class<?>) keyDeserializer.getClass())
+            .setValueDeserializer((Class<Deserializer<V>>) (Class<?>) valueDeserializer.getClass())
+            .setConsumerFactoryFn(Read.KAFKA_CONSUMER_FACTORY_FN)
+            .setConsumerConfig(config)
+            .setMaxNumRecords(Long.MAX_VALUE)
+            .build();
+  }
+
+  /**
    * Creates an uninitialized {@link Write} {@link PTransform}. Before use, Kafka configuration
    * should be set with {@link Write#withBootstrapServers(String)} and {@link Write#withTopic}
-   * along with {@link Coder}s for (optional) key and values.
+   * along with {@link Deserializer}s for (optional) key and values.
    */
   public static <K, V> Write<K, V> write() {
     return new AutoValue_KafkaIO_Write.Builder<K, V>()
         .setProducerConfig(Write.DEFAULT_PRODUCER_PROPERTIES)
         .build();
   }
 
+  /**
+   * Creates an uninitialized {@link Write} {@link PTransform}, using a Kafka {@link Serializer}
+   * based on {@link AvroCoder}. The coder writes Avro data directly without using an Avro object
+   * container.
+   */
+  @SuppressWarnings("unchecked")
+  public static <K, V> Write<K, V> toAvro(Class<K> keyClass, Class<V> valueClass) {
```

Comment:
Same comment as with `fromAvor()`. Better if this only uses public API. Also you could add 'withCoderBasedSerializers()' utility method.